src/main/java/com/suse/salt/netapi/calls/LocalCall.java

```diff
@@ -101,6 +145,76 @@ public LocalCall(String functionName, Optional<List<?>> arg,
+     * @param events the event stream to use
+     * @param cancel future to cancel the action
+     * @return a map from minion id to future of the result.
+     * @throws SaltException if anything goes wrong
+     */
+    public Map<String, CompletionStage<Result<R>>> callAsync(
+            SaltClient client,
+            Target<?> target,
+            EventStream events,
+            CompletionStage<GenericError> cancel)
+            throws SaltException {
+        LocalAsyncResult<R> lar = this.callAsync(client, target);
+        TypeToken<R> returnTypeToken = this.getReturnType();
+        Type result = ClientUtils.parameterizedType(null,
+                Result.class, returnTypeToken.getType());
+        TypeToken<Result<R>> typeToken = (TypeToken<Result<R>>) TypeToken.get(result);
+        Map<String, CompletionStage<Result<R>>> collect =
+                lar.getMinions().stream().collect(Collectors.toMap(
+                    mid -> mid,
+                        mid -> {
+                            CompletableFuture<Result<R>> future = new CompletableFuture<>();
+                            EventListener listener = new EventListener() {
+                                @Override
+                                public void notify(Event event) {
+                                    Optional<RunnerReturnEvent> runner =
+                                            RunnerReturnEvent.parse(event);
+                                    if (runner.isPresent()) {
+                                        runner.flatMap(
+                                            e -> onRunnerReturn(mid, lar.getJid()
+                                                    , e, typeToken)
+                                        ).ifPresent(future::complete);
+                                    } else {
+                                        JobReturnEvent.parse(event).flatMap(
+                                            e -> onJobReturn(mid, lar.getJid(),
+                                                    e, typeToken)
+                                        ).ifPresent(future::complete);
+                                    }
+                                }
+
+                                @Override
+                                public void eventStreamClosed(CloseReason closeReason) {
+                                    future.complete(Result.error(
+                                            new GenericError(
+                                                    "EventStream closed with reason "
+                                                            + closeReason)
+                                    ));
+                                }
+                            };
+                            events.addEventListener(listener);
+                            cancel.whenComplete((v, e) -> {
+                                if (v != null) {
+                                    future.complete(Result.error(v));
```

Comment:
If `cancel` completes exceptionally `v` will be null and `future` won't complete at all.